.hooks/pre-commit

```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+STAGED_GO_FILES=$(git diff --staged --name-only | grep "\.go$")
+
+if [[ "$STAGED_GO_FILES" = "" ]]; then
+    exit 0
+fi
+
+echo "cleaning GO dependencies"
+go mod tidy
+git add go.mod go.sum
```

Comment:
That means I get additional file into my commit where I'm not aware about it? E.g. I make some git diff, I see everything is fine, we have that and those file in the commit. Now I create the commit and afterwards I have some files in the commit which has not been listed before? Not that cool. Having files in commits should IMO be always a deliberate action.

Reply:
Yep. We also could think about just updating the file and if there is a diff cancel the commit (`exit 1`) so the user could add them himself.

Reply:
Of course there are cases were this is *not so cool*, e.g. I have dependencies added to other files which are not yet part of commit.

Reply:
So we have cases where is it no "not that cool" to add files automagically to changes. And we have cases where it is not that cool to have a check - since there are cases where the check does not make sense.

Isn't it the duty of the developer to review the diff before making a commit - or at least before pushing the commit to upstream? That is human brain-work (natural intelligence :-) )  which maybe cannot be done in a cool and satisfying way by automated checks (artificial intelligence :-)). 

> e.g. I have dependencies added to other files which are not yet part of commit.

Why should we add dependencies in an earlier commit before they are effectively used in a later commit? Wouldn't it make sense in this case to re-arrange or squash the commits?

Reply:
This was just an example how to use pre-commit hooks to make things easier. Of course it's the same like with any static check, you can have endless discussion about which check to activate and which not. But at least it's the developers choice to activate it locally. I think in some cases there will be a learning event for the developer but afterwards they will always have a clean & updated dependency definition.
Personally I will use this in one or another way to make my development workflow easier nevertheless if this is merged or not.